phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java

```diff
@@ -550,6 +550,76 @@ public BaseResultIterators(QueryPlan plan, Integer perScanLimit, Integer offset,
         }
         return ranges;
     }
+
+    private List<KeyRange> getRegionRowKeyRanges() throws SQLException {
+        List<HRegionLocation> regionLocations = getRegionBoundaries(scanGrouper);  //Load the region information
+
+        List<KeyRange> regionKeyRanges = Lists.newArrayListWithExpectedSize(regionLocations.size());
+
+        //Map each HRegionLocation to a KeyRange - no Java 8
+        for(HRegionLocation regionLocation : regionLocations){
+            HRegionInfo regionInfo = regionLocation.getRegionInfo();
+
+            //Region is upperInclusive false by definition
+            KeyRange range = KeyRange.getKeyRange(regionInfo.getStartKey(), true, regionInfo.getEndKey(), false);
+            regionKeyRanges.add(range);
+        }
+        return regionKeyRanges;
+    }
+
+    List<List<KeyRange>> getQueryRowKeyRanges(List<KeyRange> regionKeyRanges) {
+        ScanRanges scanRanges = context.getScanRanges();
+
+        RowKeySchema schema = scanRanges.getSchema();
+
+        PTable table = getTable();
+        boolean isLocalIndex = table.getIndexType() == IndexType.LOCAL;
+
+        //Use the dataplan to build the queryRowKeyRanges
+        if (isLocalIndex) {
+            // should always have a data plan when a local index is being used.
```

Comment:
The original comment says "// TODO: when implementing PHOENIX-4585, we should change this to an assert". Shall we do it now or open a Jira to track it?

Reply:
So I looked into this few weeks ago.  That JIRA, 4585, has been submitted but when I changed this to an assert I still found cases where we the dataplan was not filled in.  So I have left the original code and did not spend time investigating.